File: clojure/5/src/Problem5/bag.clj
(ns Problem5.bag
  "A bags (or multiset) is represented as map from elem to cardinality."
  (:require [clojure.set]
            [Problem5.util :refer [key-set]]))

(def bag-empty "An empty bag" {})

(defn bag-cardinality "How many elem are in the bag?" [bag elem]
  (bag elem 0))

(defn bag-add "Add one elem to the bag" [bag elem]
  (assoc bag elem (+ 1 (bag-cardinality bag elem))))

(defn bag-union "Union of any number of bags" [& bags]
  (into {} (for [elem (apply clojure.set/union (map key-set bags))]
             (let [cardinality (apply max (for [bag bags]
                                            (bag-cardinality bag elem)))]
               [elem cardinality]))))

(defn bag->list [bag]
  (for [[elem bag-cardinality] (into [] bag)
        x (repeat bag-cardinality elem)]
    x))

File: scala/src/main/scala/euler/util/bag/Bag.scala
package euler.util.bag

trait Bag[A] {

  /** How many elem are in the bag? */
  def cardinalityOf(elem: A): Int

  /** Add one elem to the bag */
  def :+(elem: A): Bag[A]

  def toSeq: Seq[A]

  def distinct: Set[A]
}

object Bag {

  def empty[A]: Bag[A] = new Bag[A] {
    override def cardinalityOf(elem: A) = 0
    override def :+(elem: A) = fromMap(Map(elem -> 1))
    override def toSeq = Seq.empty
    override def distinct = Set.empty
  }

  private def fromMap[A](map: Map[A, Int]): Bag[A] = new Bag[A] {
    override def cardinalityOf(elem: A) = map.getOrElse(elem, 0)
    override def :+(elem: A) =
      fromMap(map.updated(elem, 1 + cardinalityOf(elem)))
    override def toSeq =
      map.flatMap({
        case (elem, cardinality) => Seq.fill(cardinality)(elem)
      }).toSeq
    override def distinct = map.keySet
  }

  /** Union of any number of bags */
  def union[A](bags: Seq[Bag[A]]): Bag[A] =
    fromMap(Map[A, Int]((for {
      elem <- bags.map(_.distinct).foldLeft[Set[A]](Set.empty)(_ union _).toSeq
      cardinality = bags.map(_.cardinalityOf(elem)).max
    } yield (elem: A, cardinality: Int)): _*))
}
